test/apps/iodemo/io_demo.cc

```diff
@@ -134,76 +140,160 @@ const unsigned IoDemoRandom::_M = 0x7fffffffU;
 class P2pDemoCommon : public UcxContext {
 protected:
 
-    /* IO request header */
-    typedef struct {
-        io_op_t     op;
+    /* data validation header */
+    typedef struct __attribute__ ((packed)) {
+        uint16_t    chk_sum;
+    } chk_hdr_t;
+
+    /* transaction header */
+    typedef struct __attribute__ ((packed)) {
         uint32_t    sn;
+    } tr_hdr_t;
+
+    /* IO header */
+    typedef struct __attribute__ ((packed)) {
+        tr_hdr_t    tr;
+        io_op_t     op;
         size_t      data_size;
-    } iomsg_hdr_t;
+    } iomsg_t;
+
+    /* IO request header with check sum */
+    typedef struct __attribute__ ((packed)) {
+        chk_hdr_t   hdr;
+        iomsg_t     msg;
+    } iomsg_chk_t;
```

Comment:
why not always put checksum in iomsg? for simplicity

Reply:
did not want to increase the size and send extra bytes due to performance reason

Reply:
2 extra bytes for io message should not matter